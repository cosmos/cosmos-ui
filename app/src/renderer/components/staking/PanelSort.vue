<template lang="pug">
.panel-sort: .panel-sort-container: .sort-by(
  v-for="property in sort.properties",
  @click="orderBy(property.value, $event)",
  :class="property.class")
  .label(v-tooltip.top="property.tooltip") {{ property.title }}
</template>

<script>
export default {
  name: "panel-sort",
  methods: {
    orderBy(property) {
      let sortBys = this.$el.querySelectorAll(".sort-by")
      sortBys.forEach(el => el.classList.remove("active", "desc", "asc"))
      let index = this.sort.properties.findIndex(p => p.value === property)
      let el = sortBys[index]

      if (this.sort.property === property) {
        if (this.sort.order === "asc") {
          this.sort.order = "desc"
        } else {
          this.sort.order = "asc"
        }
      } else {
        this.sort.property = property
      }
      if (this.sort.order === "asc") {
        el.classList.add("asc")
      } else {
        el.classList.add("desc")
      }
      el.classList.add("active")
    }
  },
  props: ["sort"]
}
</script>

<style lang="stylus">
@require '~variables'

.panel-sort-container
  display flex
  height 2rem
  border-bottom px solid var(--bc)

.sort-by
  flex 1
  cursor pointer
  user-select none
  display flex
  align-items center
  position relative
  min-width 0

  .label
    color var(--link)
    white-space nowrap
    text-overflow ellipsis
    overflow hidden

  &:after
    display block
    font-family FontAwesome
    color var(--dim)
    padding-left 0.25rem

  &.asc:after
    content '\f0d8'

  &.desc:after
    content '\f0d7'

  &.active
    .label
      color var(--tertiary)
    &:after
      color var(--tertiary)

  &.name
    flex 2

  &.action
    flex 0.5

  &.hidden
    visibility hidden

@media screen and (max-width: 768px)
  .sort-by
    &.id
      display none
</style>
