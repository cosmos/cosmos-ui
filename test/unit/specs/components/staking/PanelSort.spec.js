import setup from "../../../helpers/vuex-setup"
import PanelSort from "renderer/components/staking/PanelSort"

describe(`PanelSort`, () => {
  let wrapper
  let instance = setup()

  beforeEach(() => {
    let test = instance.mount(PanelSort, {
      propsData: {
        sort: {
          order: `asc`
        },
        properties: [
          {
            value: `id`,
            title: `ID`
          },
          {
            value: `amount`,
            title: `AMOUNT`
          }
        ]
      }
    })
    wrapper = test.wrapper
  })

  it(`has the expected html structure`, () => {
    expect(wrapper.vm.$el).toMatchSnapshot()
  })

  it(`should show an arrow according to ordering`, () => {
    let firstCol = wrapper.vm.$el.querySelector(`.sort-by`)
    let link = wrapper.vm.$el.querySelector(`.sort-by-link`)
    expect(firstCol.className.split(` `)).not.toContain(`asc`)
    expect(firstCol.className.split(` `)).not.toContain(`desc`)
    link.click()
    expect(firstCol.className.split(` `)).not.toContain(`desc`)
    expect(firstCol.className.split(` `)).toContain(`asc`)
    link.click()
    expect(firstCol.className.split(` `)).not.toContain(`asc`)
    expect(firstCol.className.split(` `)).toContain(`desc`)
    link.click()
    expect(firstCol.className.split(` `)).toContain(`asc`)
    expect(firstCol.className.split(` `)).not.toContain(`desc`)
  })

  it(`should only sort one col actively`, () => {
    let firstCol = wrapper.vm.$el.querySelectorAll(`.sort-by`)[0]
    let secondCol = wrapper.vm.$el.querySelectorAll(`.sort-by`)[1]
    let firstLink = wrapper.vm.$el.querySelectorAll(`.sort-by-link`)[0]
    let secondLink = wrapper.vm.$el.querySelectorAll(`.sort-by-link`)[1]
    firstLink.click()
    secondLink.click()
    expect(firstCol.className.split(` `)).not.toContain(`asc`)
    expect(secondCol.className.split(` `)).toContain(`asc`)
  })
})
