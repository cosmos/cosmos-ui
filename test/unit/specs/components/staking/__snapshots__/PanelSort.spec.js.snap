// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PanelSort has the expected html structure 1`] = `
<tr
  class="panel-sort-container"
>
  <th
    class="sort-by"
  >
    <a
      class="sort-by-link"
    >
      
      ID
    
    </a>
  </th>
  <th
    class="sort-by"
  >
    <a
      class="sort-by-link"
    >
      
      AMOUNT
    
    </a>
  </th>
</tr>
`;
